### PR TITLE
perf(picker): Fetch Shared Link info on select

### PR DESCRIPTION
**Pull request by ConradJChan**
This removes the shared link fields from the initial folder request for content picker. If `canSetShareAccess` prop is true, when an item is selected, then an API call to fetch the missing shared link fields will be invoked. Also if no shared link currently exists for the item, then one will be created with enterprise defaults.

Content Picker

- [ ] Unit tests

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @ConradJChan has signed the CLA. Thanks for the pull request!

File: src/elements/content-picker/ContentPicker.js [hunk 1]
```diff
@@ -820,6 +827,55 @@ class ContentPicker extends Component<Props, State> {
         });
     };
 
+    /**
+     * Show the inline shared link dropdown
+     * @param {BoxItem} item - The item (folder, file, weblink)
+     * @returns {Promise}
+     */
+    showSharedLinkDropdown = (item: BoxItem): void => {
```
Comment (jeremypress) [inline src/elements/content-picker/ContentPicker.js]: rename? This method seems to fetch info but doesn't actually show anything.

File: src/elements/content-picker/shareAccessCellRenderer.js [hunk 2]
```diff
@@ -15,11 +16,16 @@ export default (
...
     hasHitSelectionLimit: boolean,
 ) => ({ rowData }: { rowData: BoxItem }) => {
-    if (!isRowSelectable(selectableType, extensionsWhitelist, hasHitSelectionLimit, rowData)) {
+    if (!isRowSelectable(selectableType, extensionsWhitelist, hasHitSelectionLimit, rowData) || !rowData.selected) {
         return <span />;
     }
 
-    return (
+    const { allowed_shared_link_access_levels } = rowData;
+    const isLoading = !allowed_shared_link_access_levels;
```
Comment (jeremypress) [inline src/elements/content-picker/shareAccessCellRenderer.js]: Minor, but do we need a failed state here now that it's a separate call?

File: src/elements/content-picker/shareAccessCellRenderer.js [hunk 3]
```diff
@@ -15,11 +16,16 @@ export default (
...
+
+    return isLoading ? (
+        <LoadingIndicator className="bcp-ShareAccessSelect--loading" />
```
Comment (jstoffan) [inline src/elements/content-picker/shareAccessCellRenderer.js]: Modifiers should be used together with their base class, like `className="bcp-ShareAccessSelect bcp-ShareAccessSelect--loading"`. Also, since this isn't actually an instance of the ShareAccessSelect component, the class should probably be renamed.

File: src/elements/content-picker/ContentPicker.js [hunk 4]
```diff
@@ -820,6 +827,55 @@ class ContentPicker extends Component<Props, State> {
         });
     };
 
+    /**
+     * Show the inline shared link dropdown
+     * @param {BoxItem} item - The item (folder, file, weblink)
+     * @returns {Promise}
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Looks like the JSDoc and flowtypes don't match for the return value.

File: src/elements/content-picker/ContentPicker.js [hunk 5]
```diff
@@ -759,8 +761,8 @@ class ContentPicker extends Component<Props, State> {
      * @param {Object} item file or folder object
      * @return {void}
      */
-    select = (item: BoxItem): void => {
-        const { type: selectableType, maxSelectable }: Props = this.props;
+    select = (item: BoxItem, forceSharedLink: boolean = true): void => {
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Can we make the latter argument an options object so we can use named parameters? Otherwise it's difficult to tell what passing `false` does here by looking at only the calling code.

File: src/elements/content-picker/ContentPicker.js [hunk 6]
```diff
@@ -795,7 +797,7 @@ class ContentPicker extends Component<Props, State> {
 
             // Check if we hit the selection limit
             // Ignore when in single file selection mode.
-            if (hasHitSelectionLimit && !isSingleFileSelection) {
+            if (hasHitSelectionLimit && !isSingleFileSelection && !selectedKeys.indexOf(cacheKey)) {
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: This new condition seems unrelated to the selection limit. Can we update the comment or break it out into a variable so it's clear what we want to test?

File: src/elements/content-picker/ContentPicker.js [hunk 7]
```diff
@@ -820,6 +827,55 @@ class ContentPicker extends Component<Props, State> {
...
+     * Show the inline shared link dropdown
+     * @param {BoxItem} item - The item (folder, file, weblink)
+     * @returns {Promise}
+     */
+    showSharedLinkDropdown = (item: BoxItem): void => {
+        const { id, type }: BoxItem = item;
+
+        switch (type) {
+            case TYPE_FOLDER:
+                this.api.getFolderAPI().getFolderFields(id, this.handleSharedLinkResponse, () => {}, {
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Is swallowing the error response desirable? If so, can we use `lodash/noop`, instead?

File: src/elements/content-picker/ContentPicker.js [hunk 8]
```diff
@@ -820,6 +827,55 @@ class ContentPicker extends Component<Props, State> {
...
+    };
+
+    /**
+     * Handles the shared link info by either creating a share link using enterprise defaults if
+     * it does not already exist, otherwise update the item in the state currentCollection.
+     *
+     * @param {Object} item file or folder
+     * @returns {void}
+     */
+    handleSharedLinkResponse = (item: BoxItem) => {
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: `handleSharedLinkSuccess`?

File: src/elements/content-picker/ContentPicker.js [hunk 9]
```diff
@@ -820,6 +827,55 @@ class ContentPicker extends Component<Props, State> {
...
+        // if no shared link currently exists, create a shared link with enterprise default
+        if (!item[FIELD_SHARED_LINK]) {
+            this.changeShareAccess(null, item);
+        } else {
+            const { selected } = this.state;
+            const { id, type } = item;
+            const cacheKey = this.api.getAPI(type).getCacheKey(id);
+            // if shared link already exists, update the collection in state
+            this.updateItemInCollection(item);
+            if (item.selected && item !== selected[cacheKey]) {
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Does this handle the case where the user selects a new item before the shared link response returns? May want to add a comment.

File: src/elements/content-picker/ContentPicker.js [hunk 10]
```diff
@@ -845,13 +901,31 @@ class ContentPicker extends Component<Props, State> {
...
 
+    /**
+     * Updates the BoxItem in the state's currentCollection
+     *
+     * @param {Object} item file or folder object
+     * @returns {void}
+     */
+    updateItemInCollection = (item: BoxItem) => {
+        const { currentCollection } = this.state;
+        const { items = [] } = currentCollection;
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Can `items` be undefined here?

File: src/api/Folder.js [hunk 11]
```diff
@@ -160,7 +161,7 @@ class Folder extends Item {
             new FileAPI(this.options),
             new WebLinkAPI(this.options),
         );
-        this.itemCache = (this.itemCache || []).concat(flattened);
+        this.itemCache = union(this.itemCache, flattened);
```
[the remark this answers is not in the document]
Comment (ConradJChan) [reply src/api/Folder.js]: Yeah I can most likely revert this now, i think it was only an issue when I was previously fetching the folder again for the shared link info and it would duplicate the entries

File: src/elements/content-picker/shareAccessCellRenderer.js (hunk 3 above)
Comment (ConradJChan) [reply src/elements/content-picker/shareAccessCellRenderer.js]: Makes sense, how's `bcp-share-access-loading`?

File: src/elements/content-picker/ContentPicker.js (hunk 6 above)
Comment (ConradJChan) [reply src/elements/content-picker/ContentPicker.js]: I'll do both thanks

File: src/elements/content-picker/ContentPicker.js (hunk 9 above)
Comment (ConradJChan) [reply src/elements/content-picker/ContentPicker.js]: Yeah you're right, if the user clicks another item before the XHR completes it gets cancelled. I'll add a comment about that

File: src/elements/content-picker/ContentPicker.js (hunk 10 above)
Comment (ConradJChan) [reply src/elements/content-picker/ContentPicker.js]: It's unlikely that `items` could be undefined here, do you see a potential issue?

File: test/integration/content-picker/ContentPicker.e2e.test.js [hunk 12]
```diff
@@ -11,51 +13,89 @@ const helpers = {
...
+        );
+
+        ['319004423111', '308566419514', '308409990441'].forEach(fileId => {
+            cy.route('GET', `**/files/${fileId}?fields=allowed_shared_link_access_levels,shared_link`, {
+                ...getSharedLinkRsp,
+                id: fileId,
+            });
+
+            cy.route('PUT', `**/files/${fileId}`, {
+                ...createSharedLinkRsp,
```
Comment (jstoffan) [inline test/integration/content-picker/ContentPicker.e2e.test.js]: I think we want to avoid explicit imports when possible. Can we import these via `cy.fixture`, instead?

File: test/integration/content-picker/ContentPicker.e2e.test.js [hunk 13]
```diff
@@ -11,51 +13,89 @@ const helpers = {
...
+    },
 };
 
 describe('ContentPicker', () => {
+    beforeEach(() => {
+        cy.server();
+
+        cy.route(
+            'GET',
+            '**/folders/51964781421?fields=id,name,type,size,parent,extension,permissions,path_collection,modified_at,created_at,modified_by,has_collaborations,is_externally_owned,item_collection&direction=asc&limit=50&offset=0&sort=name',
```
Comment (jstoffan) [inline test/integration/content-picker/ContentPicker.e2e.test.js]: Can we avoid hardcoding the file/folder ids? Otherwise it seems like the tests can be pretty brittle if the fixture data changes.

File: test/integration/content-picker/ContentPicker.e2e.test.js [hunk 14]
```diff
@@ -11,51 +13,89 @@ const helpers = {
...
+    },
+    unselectRow(rowNum) {
+        this.getRow(rowNum)
+            .as('row')
             .click()
             .find('input[type="checkbox"]')
-            .should('not.be.checked'),
+            .should('not.be.checked');
+
+        cy.get('@row')
```
Comment (jstoffan) [inline test/integration/content-picker/ContentPicker.e2e.test.js]: Should this helper also return `cy.get('@row')`?

File: src/api/Folder.js [hunk 15]
```diff
@@ -174,33 +174,85 @@ class Folder extends Item {
...
+     * @param {Array<String>} fields Array of field strings
+     * @return {Promise}
      */
-    folderRequest(): Promise<void> {
+    folderRequest(options: FetchOptions = {}, successHandler?: Function = this.folderSuccessHandler): Promise<any> {
         if (this.isDestroyed()) {
             return Promise.reject();
         }
 
+        const { fields, noPagination } = options;
```
Comment (jstoffan) [inline src/api/Folder.js]: These could be destructured in the signature.

File: src/elements/content-picker/ContentPicker.js [hunk 16]
```diff
@@ -757,10 +759,11 @@ class ContentPicker extends Component<Props, State> {
      *
      * @private
      * @param {Object} item file or folder object
+     * @param {boolean} options.forceSharedLink Force a shared link if no link exists
      * @return {void}
      */
-    select = (item: BoxItem): void => {
-        const { type: selectableType, maxSelectable }: Props = this.props;
+    select = (item: BoxItem, options: Object = {}): void => {
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Can we use something other than the `Object` type for `options`?

File: src/elements/content-picker/ContentPicker.js [hunk 17]
```diff
@@ -779,6 +782,10 @@ class ContentPicker extends Component<Props, State> {
         const cacheKey: string = this.api.getAPI(type).getCacheKey(id);
         const existing: BoxItem = selected[cacheKey];
         const existingFromCache: BoxItem = this.api.getCache().get(cacheKey);
+        const { forceSharedLink = true } = options;
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: These can be destructured in the function signature.

File: src/elements/content-picker/ContentPicker.js [hunk 18]
```diff
@@ -779,6 +782,10 @@ class ContentPicker extends Component<Props, State> {
...
+        const existInSelected = selectedKeys.indexOf(cacheKey) !== -1;
+        const { permissions = {} }: BoxItem = existingFromCache;
```
Comment (jstoffan) [inline src/elements/content-picker/ContentPicker.js]: Should this permissions object come from the passed-in `item`, instead?

File: src/elements/content-picker/ContentPicker.js (hunk 9 above)
Comment (jstoffan) [reply src/elements/content-picker/ContentPicker.js]: Consider adding a comment.

File: src/elements/content-picker/shareAccessCellRenderer.js [hunk 19]
```diff
@@ -15,11 +16,24 @@ export default (
     extensionsWhitelist: string[],
     hasHitSelectionLimit: boolean,
 ) => ({ rowData }: { rowData: BoxItem }) => {
-    if (!isRowSelectable(selectableType, extensionsWhitelist, hasHitSelectionLimit, rowData)) {
+    const { permissions = {} }: BoxItem = rowData;
+    const { can_set_share_access: rowCanSetShareAccess }: BoxItemPermission = permissions;
```
Comment (jstoffan) [inline src/elements/content-picker/shareAccessCellRenderer.js]: Should this be `itemCanSetShareAccess` to match the similar usage in ContentPicker? You could also consider using `lodash/get` to cut down on the multiple destructures/default values in both spots: 

```
const itemCanSetShareAccess = getProp(rowData, 'permissions.can_set_share_access', false);
```

File: src/elements/content-picker/shareAccessCellRenderer.js (hunk 2 above)
Comment (jstoffan) [reply src/elements/content-picker/shareAccessCellRenderer.js]: @ConradJChan, thoughts on @jeremypress's comment above?

**Comment by jstoffan** [comment]
May also want to consider using the `perf` type for the commit message, as well.

File: src/elements/content-picker/shareAccessCellRenderer.js (hunk 2 above)
Comment (ConradJChan) [reply src/elements/content-picker/shareAccessCellRenderer.js]: Yeah I checked with Zach and Rohit and since it seems like an uncommon error, just leaving it as is for now
